### PR TITLE
groth16-gpu

### LINKED ISSUE #129: GPU frankin crypto
* issue 
https://github.com/0xEigenLabs/eigen-zkvm/pull/128#discussion_r1365673007

### PR DESCRIPTION
**Pull request by ibmp33**
fix #129

### CODE REVIEW

File: groth16/src/lib.rs [hunk 1]
```diff
@@ -1,8 +1,29 @@
...
+    pub mod r1cs_file;
+    pub mod reader;
+    
+    pub use bellperson;
+// }
+
+// #[cfg(any(feature = "cuda", feature = "opencl"))]
+// pub use gpu_specific::*;
+
+// #[cfg(not(any(feature = "cuda", feature = "opencl")))]
```
Comment (ibmp33) [inline groth16/src/lib.rs:21]: Is the franklin_crypto project still being used in the non_gpu scenario?

File: plonky/Cargo.toml [hunk 2]
```diff
@@ -46,7 +46,7 @@ ark-std = { version = "0.4.0", optional = true }
 env_logger = "0.10"
 
 [features]
-default = ["bellman_vk_codegen/multicore", "franklin-crypto/multicore", "ethabi", "recursive_aggregation_circuit/multicore"]
+default = ["bellman_vk_codegen/multicore", "franklin-crypto/multicore", "ethabi", "recursive_aggregation_circuit/multicore", "algebraic/default"]
```
Comment (ibmp33) [inline plonky/Cargo.toml:49]: cargo build --release --features "opencl plonky/default recursion/default" # This command cannot include the default algebraic feature

File: groth16/Cargo.toml [hunk 3]
```diff
@@ -3,35 +3,42 @@ name = "groth16"
...
-profiler_macro = {git = "https://github.com/ChengYueJia/profiler-rs", rev = "6d6a77f8"}
+profiler_macro = { git = "https://github.com/ChengYueJia/profiler-rs", rev = "6d6a77f8" }
 ark-std = { version = "0.4.0", optional = true }
+algebraic = { path = "../algebraic", default-features = false }
+
+ff = { version = "0.13.0", features = ["derive"], optional = true }
+pairing = { version = "0.23.0", optional = true }
+group = { version = "0.13.0", optional = true }
+blstrs = { version = "0.7.0", features = ["__private_bench"], optional = true }
+bellperson = { version = "0.25", default-features = false, optional = true }
```
Comment (eigmax) [inline groth16/Cargo.toml:29]: Use frankin-crypto by default.

File: plonky/Cargo.toml (hunk 2 above)
Comment (eigmax) [reply plonky/Cargo.toml:49]: --features "A feature" --features "B features" ...

File: algebraic/src/r1cs_file.rs [hunk 4]
```diff
@@ -163,6 +232,30 @@ fn read_custom_gates_list<R: Read, E: ScalarEngine>(
...
+            custom_gate
+                .parameters
+                .push(read_field::<&mut R, E>(&mut reader)?);
+        }
+        custom_gates.push(custom_gate);
+    }
+    Ok(custom_gates)
+}
+
 fn read_custom_gates_uses_list<R: Read>(
```
Comment (eigmax) [inline algebraic/src/r1cs_file.rs:259]: Do we miss a feature config here?

File: groth16/Cargo.toml [hunk 5]
```diff
@@ -3,35 +3,43 @@ name = "groth16"
...
+profiler_macro = { git = "https://github.com/ChengYueJia/profiler-rs", rev = "6d6a77f8" }
 ark-std = { version = "0.4.0", optional = true }
+algebraic = { path = "../algebraic", default-features = false, optional = true }
+algebraic-gpu = { path = "../algebraic-gpu", default-features = false, optional = true }
+
+ff = { version = "0.13.0", features = ["derive"], optional = true }
+pairing = { version = "0.23.0", optional = true }
+group = { version = "0.13.0", optional = true }
+blstrs = { version = "0.7.0", features = ["__private_bench"], optional = true }
+bellperson = { version = "0.26", default-features = false, features = ["groth16"], optional = true }
```
Comment (eigmax) [inline groth16/Cargo.toml:30]: can we group up the gpu dependencies from the cpu dependencies?

File: algebraic-gpu/test-vectors/mycircuit-input1.json [hunk 6]
```diff
@@ -0,0 +1,2 @@
+{"a": 3, "b": 11}
```
Comment (eigmax) [inline algebraic-gpu/test-vectors/mycircuit-input1.json:1]: Can we hardcode those files in the source code accordingly to reduce the file scale?  same for the algebraic project

File: algebraic-gpu/test-vectors/mycircuit.circom [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+template Multiplier() {
+    signal input a;
+    signal input b;
+    signal output c;
+
+    c <== a*b;
+}
+
+component main = Multiplier();
+
```
Comment (eigmax) [inline algebraic-gpu/test-vectors/mycircuit.circom:11]: Any way to reuse the circom and wasm files for both cpu and gpu algebraic?

File: groth16/src/r1cs_file.rs [hunk 8]
```diff
@@ -0,0 +1,414 @@
+// some codes borrowed from https://github.com/poma/zkutil/blob/master/src/r1cs_reader.rs
```
Comment (eigmax) [inline groth16/src/r1cs_file.rs:1]: Can we remove this file? since we have one in the algebraic-gpu

**Pull request merged**